### PR TITLE
Initial work

### PR DESCRIPTION
**Pull request by Filco306**
Some things added here. 

- Started on the training functions, epoch runnings etc. 
- Fixed two dataloader functions. 
- Built a VAE, (very) much inspired by another implementation for training and to more deeply understand the implementation of a vanilla VAE. It should be trainable in the coming few days. 
- Built some tests for some functions.

### CODE REVIEW

**Comment by Filco306** [comment]
Not done yet lol, brb with more features

**Comment by Filco306** [comment]
Opened this up again, ready for inspection. I am able to run the model, but the loss is suspiciously varying, which makes me wonder if there is something wrong somewhere. I have not gotten to fixing it, but there probably is something somewhere that might be a bit off.

**Comment by Filco306** [comment]
@TobiasNorlund

File: src/vae.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        x_hat = x_hat.view(-1, self.q_in)
+        mu, log_sigma = self.q_mu(x_hat), self.q_vars(x_hat)
+        return mu, log_sigma
+
+    def z(self, mu, log_sigma):
+        # Sample from z
+        sigma = log_sigma.mul(0.5).exp_()
+        # Here we sample from a normal distribution
+        #
+        ret = torch.randn(mu.shape)
```
Comment (TobiasNorlund) [inline src/vae.py:173]: How are the gradients able to propagate through this step?

Comment (Filco306) [reply src/vae.py:173]: It is not supposed to, but rather through `sigma` and `mu` - it is the [reparametrization trick](https://stats.stackexchange.com/questions/199605/how-does-the-reparameterization-trick-for-vaes-work-and-why-is-it-important)

Comment (TobiasNorlund) [reply src/vae.py:173]: Ah yeah they flow through the `mul` and `add_` below :) thanks!

**Review by TobiasNorlund — APPROVED** [review]
Very good! LGTM! :clap:
